modal/_utils/function_utils.py

```diff
@@ -190,6 +190,42 @@ def serialized_function(self) -> bytes:
             logger.debug(f"Serializing function for class service function {self.cls.__qualname__} as empty")
             return b""
 
+    def get_cls_vars(self) -> Dict[str, Any]:
+        if self.cls is not None:
+            cls_vars = {
+                attr: getattr(self.cls, attr)
+                for attr in dir(self.cls)
+                if not callable(getattr(self.cls, attr)) and not attr.startswith("__")
+            }
+            return cls_vars
+        return {}
+
+    def get_cls_var_attrs(self) -> Dict[str, Any]:
+        import dis
+
+        import opcode
+
+        LOAD_ATTR = opcode.opmap["LOAD_ATTR"]
+        STORE_ATTR = opcode.opmap["STORE_ATTR"]
+
+        func = self.raw_f
+        code = func.__code__
+        f_attr_ops = set()
+        for instr in dis.get_instructions(code):
+            if instr.opcode == LOAD_ATTR:
+                f_attr_ops.add(instr.argval)
+            elif instr.opcode == STORE_ATTR:
+                f_attr_ops.add(instr.argval)
+                logger.warning(
+                    "\n\nWarning: %s set in @build. You must set class variables in "
```

Comment:
@mwaskom does this warning seem ok?

Reply:
I think we'd probably want to issue it via the `warning` module — but since it's not directly related let's defer this change and think about it a tiny bit to make sure it's a good idea. It makes sense, but could potentially be annoying if we don't do it right.

Reply:
Ok I'll remove it for now. Are we otherwise good to merge? This change will force rebuilds for all users that have static variables in their class so I do want to make sure things are ok on that front

Reply:
> This change will force rebuilds for all users that have static variables in their class so I do want to make sure things are ok on that front.

Ah that's a great callout — we should be a little conservative about that.

On the one hand, layer defined by the class `@build` function should be the final layer(s) so won't trigger cascading  reinstalls of packages, which is often where people get into trouble (if they don't pin their dependency versions). On the other hand, they're often used for doing things like downloading models, which will take a while (and in some rare cases, break, if people point at a model revision that gets removed).

I guess it should specifically be limited to apps with build functions *that read or write to class variables*, is that right? In that case, I suspect it's a narrow-enough population who will directly benefit going forward.

Reply:
This is correct. It's only apps with build functions that read / write to class variables. I don't think that's a large population, but good to flag in any case.